File: python/listas/lista-circular.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


head = None


def menu():
    print("\n********** Menú principal **********")
    print("1. Insertar al principio")
    print("2. Insertar al final")
    print("3. Insertar en posición")
    print("4. Eliminar del principio")
    print("5. Eliminar del final")
    print("6. Eliminar nodo después de la ubicación")
    print("7. Buscar un elemento")
    print("8. Mostrar")
    print("9. Salir")

    choice = int(input("\nIngrese su opción: "))

    if choice == 1:
        beg_insert()
    elif choice == 2:
        last_insert()
    elif choice == 3:
        select_insert()
    elif choice == 4:
        begin_delete()
    elif choice == 5:
        last_delete()
    elif choice == 6:
        select_delete()
    elif choice == 7:
        search()
    elif choice == 8:
        display()
    elif choice == 9:
        print("Saliendo...")
        exit()
    else:
        print("Opción no válida")


# 1. Insertar al inicio
def beg_insert():
    global head
    item = int(input("Ingrese valor: "))
    new_node = Node(item)

    if head is None:
        head = new_node
        new_node.next = head
    else:
        temp = head
        while temp.next != head:
            temp = temp.next

        new_node.next = head
        temp.next = new_node
        head = new_node

    print("Nodo insertado al inicio")


# 2. Insertar al final
def last_insert():
    global head
    item = int(input("Ingrese valor: "))
    new_node = Node(item)

    if head is None:
        head = new_node
        new_node.next = head
    else:
        temp = head
        while temp.next != head:
            temp = temp.next

        temp.next = new_node
        new_node.next = head

    print("Nodo insertado al final")


# 3. Insertar después de una ubicación
def select_insert():
    global head
    item = int(input("Ingrese el valor a insertar: "))
    loc = int(input("Ingrese la posición después de la cual insertar: "))

    if head is None or loc <= 0:
        print("No se puede insertar")
        return

    new_node = Node(item)
    temp = head

    for i in range(1, loc):
        temp = temp.next
        if temp == head:
            print("Ubicación fuera de rango")
            return

    new_node.next = temp.next
    temp.next = new_node

    print("Nodo insertado")


# 4. Eliminar al inicio
def begin_delete():
    global head

    if head is None:
        print("Lista vacía")
        return

    if head.next == head:
        head = None
    else:
        temp = head
        while temp.next != head:
            temp = temp.next

        head = head.next
        temp.next = head

    print("Primer nodo eliminado")


# 5. Eliminar al final
def last_delete():
    global head

    if head is None:
        print("Lista vacía")
        return

    if head.next == head:
        head = None
        print("Se eliminó el único nodo de la lista")
        return

    prev = None
    ptr = head

    while ptr.next != head:
        prev = ptr
        ptr = ptr.next

    prev.next = head

    print("Último nodo eliminado")


# 6. Eliminar después de una posición
def select_delete():
    global head

    loc = int(input("Ingrese posición después de la cual desea eliminar: "))

    if head is None or loc <= 0:
        print("Lista vacía o posición inválida")
        return

    ptr1 = head
    ptr = head.next

    for i in range(2, loc):
        ptr1 = ptr
        ptr = ptr.next

        if ptr == head:
            print("Ubicación fuera de los límites")
            return

    if ptr == head:
        print("Ubicación fuera de los límites")
        return

    ptr1.next = ptr.next
    print(f"Nodo eliminado en la posición {loc + 1}")


# 7. Buscar elemento
def search():
    global head

    if head is None:
        print("Lista vacía")
        return

    item = int(input("Ingrese el elemento a buscar: "))

    ptr = head
    i = 1
    found = False

    while True:
        if ptr.data == item:
            print(f"Elemento encontrado en la posición {i}")
            found = True

        ptr = ptr.next
        i += 1

        if ptr == head:
            break

    if not found:
        print("Elemento no encontrado")


# 8. Mostrar lista
def display():
    global head

    if head is None:
        print("Nada que imprimir")
        return

    ptr = head
    print("\nMostrando lista circular:")

    while True:
        print(ptr.data)
        ptr = ptr.next

        if ptr == head:
            break


# Programa principal
while True:
    menu()

File: javascript/listas/lista-circular.js
const readline = require('readline');

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

let head = null;

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function menu() {
    console.log("\n********** Menú principal **********");
    console.log("1. Insertar al principio");
    console.log("2. Insertar al final");
    console.log("3. Insertar en posición");
    console.log("4. Eliminar del principio");
    console.log("5. Eliminar del final");
    console.log("6. Eliminar nodo después de la ubicación");
    console.log("7. Buscar un elemento");
    console.log("8. Mostrar");
    console.log("9. Salir");

    rl.question("\nIngrese su opción: ", answer => {
        const choice = parseInt(answer);

        switch (choice) {
            case 1: begInsert(); break;
            case 2: lastInsert(); break;
            case 3: selectInsert(); break;
            case 4: beginDelete(); break;
            case 5: lastDelete(); break;
            case 6: selectDelete(); break;
            case 7: search(); break;
            case 8: display(); break;
            case 9:
                console.log("Saliendo...");
                rl.close();
                return;
            default:
                console.log("Opción no válida");
                menu();
        }
    });
}

// 1. Insertar al inicio
function begInsert() {
    rl.question("Ingrese valor: ", value => {
        const item = parseInt(value);
        const newNode = new Node(item);

        if (head === null) {
            head = newNode;
            newNode.next = head;
        } else {
            let temp = head;
            while (temp.next !== head) {
                temp = temp.next;
            }
            newNode.next = head;
            temp.next = newNode;
            head = newNode;
        }

        console.log("Nodo insertado al inicio");
        menu();
    });
}

// 2. Insertar al final
function lastInsert() {
    rl.question("Ingrese valor: ", value => {
        const item = parseInt(value);
        const newNode = new Node(item);

        if (head === null) {
            head = newNode;
            newNode.next = head;
        } else {
            let temp = head;
            while (temp.next !== head) {
                temp = temp.next;
            }
            temp.next = newNode;
            newNode.next = head;
        }

        console.log("Nodo insertado al final");
        menu();
    });
}

// 3. Insertar después de una ubicación
function selectInsert() {
    rl.question("Ingrese el valor a insertar: ", value => {
        const item = parseInt(value);
        const newNode = new Node(item);

        rl.question("Ingrese la posición después de la cual insertar: ", locStr => {
            const loc = parseInt(locStr);

            if (head === null || loc <= 0) {
                console.log("No se puede insertar");
                menu();
                return;
            }

            let temp = head;

            for (let i = 1; i < loc; i++) {
                temp = temp.next;
                if (temp === head) {
                    console.log("Ubicación fuera de rango");
                    menu();
                    return;
                }
            }

            newNode.next = temp.next;
            temp.next = newNode;

            console.log("Nodo insertado");
            menu();
        });
    });
}

// 4. Eliminar al inicio
function beginDelete() {
    if (head === null) {
        console.log("Lista vacía");
        menu();
        return;
    }

    if (head.next === head) {
        head = null;
    } else {
        let temp = head;
        while (temp.next !== head) {
            temp = temp.next;
        }
        head = head.next;
        temp.next = head;
    }

    console.log("Primer nodo eliminado");
    menu();
}

// 5. Eliminar al final
function lastDelete() {
    if (head === null) {
        console.log("Lista vacía");
        menu();
        return;
    }

    if (head.next === head) {
        head = null;
        console.log("Se eliminó el único nodo de la lista");
        menu();
        return;
    }

    let ptr = head;
    let prev = null;

    while (ptr.next !== head) {
        prev = ptr;
        ptr = ptr.next;
    }

    prev.next = head;
    console.log("Último nodo eliminado");
    menu();
}

// 6. Eliminar después de una posición
function selectDelete() {
    rl.question("Ingrese posición después de la cual desea eliminar: ", locStr => {
        const loc = parseInt(locStr);

        if (head === null || loc <= 0) {
            console.log("Lista vacía o posición inválida");
            menu();
            return;
        }

        let ptr1 = head;
        let ptr = head.next;

        for (let i = 2; i < loc; i++) {
            ptr1 = ptr;
            ptr = ptr.next;

            if (ptr === head) {
                console.log("Ubicación fuera de los límites");
                menu();
                return;
            }
        }

        if (ptr === head) {
            console.log("Ubicación fuera de los límites");
            menu();
            return;
        }

        ptr1.next = ptr.next;
        console.log(`Nodo eliminado en la posición ${loc + 1}`);
        menu();
    });
}

// 7. Buscar elemento
function search() {
    if (head === null) {
        console.log("Lista vacía");
        menu();
        return;
    }

    rl.question("Ingrese el elemento a buscar: ", val => {
        const item = parseInt(val);

        let ptr = head;
        let i = 1;
        let found = false;

        do {
            if (ptr.data === item) {
                console.log(`Elemento encontrado en la posición ${i}`);
                found = true;
            }
            ptr = ptr.next;
            i++;
        } while (ptr !== head);

        if (!found) {
            console.log("Elemento no encontrado");
        }

        menu();
    });
}

// 8. Mostrar lista
function display() {
    if (head === null) {
        console.log("Nada que imprimir");
        menu();
        return;
    }

    let ptr = head;

    console.log("\nMostrando lista circular:");
    do {
        console.log(ptr.data);
        ptr = ptr.next;
    } while (ptr !== head);

    menu();
}

// Iniciar programa
menu();
